src/software/ai/hl/stp/tactic/goalie/goalie_tactic_test.py

```diff
@@ -0,0 +1,135 @@
+        # ball slow inside friendly defense area
+        (geom.Point(-4, 0.8), geom.Vector(-0.2, 0), geom.Point(0, 2)),
+        # ball slow inside friendly defense area
+        (geom.Point(-4, 0.8), geom.Vector(-0.2, 0), geom.Point(-4, 0)),
+        # ball stationary inside friendly defense area
+        (geom.Point(-4, 0.0), geom.Vector(0.0, 0), geom.Field().friendlyGoalpostPos()),
+        # ball stationary inside no-chip rectangle
+        (
+            geom.Field().friendlyGoalCenter() + geom.Vector(0.1, 0.1),
+            geom.Vector(-0.2, 0),
+            geom.Point(-4, -1),
+        ),
+        # ball fast inside no-chip rectangle but no intersection with goal
+        (
+            geom.Field().friendlyGoalCenter() + geom.Vector(0.1, 0),
+            geom.Vector(0, -0.5),
+            geom.Point(-3.5, 1),
+        ),
+        # ball moving out from inside defense area
+        (
+            geom.Field().friendlyGoalCenter() + geom.Vector(0.5, 0),
+            geom.Vector(0.5, 0),
+            geom.Point(-3.5, 0),
+        ),
+        # ball slow inside no-chip rectangle
+        (
+            geom.Field().friendlyGoalCenter() + geom.Vector(0.1, 0),
+            geom.Vector(0.1, -0.1),
+            geom.Point(-3.5, 1),
+        ),
+        # TODO (#2167): This test fails so disabling for Robocup
+        # ball moving into goal from inside defense area
+        (
+            geom.Field().friendlyGoalCenter() + geom.Vector(0.5, 0),
+            geom.Vector(-0.5, 0),
+            geom.Point(-3.5, 0),
+        ),
+        # TODO (#2167): This test fails so disabling for Robocup
+        # ball moving up and out of defense area
+        (
+            geom.Field().friendlyGoalCenter() + geom.Vector(0.3, 0),
+            geom.Vector(0, 1),
+            geom.Point(-3.5, 0),
+        ),
+        # TODO (#2167): This test fails so disabling for Robocup
+        # ball moving down and out goal from defense area
+        (
+            geom.Field().friendlyGoalCenter() + geom.Vector(0.3, 0),
+            geom.Vector(0, -0.7),
+            geom.Point(-3.5, 0),
+        ),
+    ],
+)
+def test_goalie_blocks_shot(
+    ball_initial_position, ball_initial_velocity, robot_initial_position, tactic_runner
+):
+    # Setup Robot
+    tactic_runner.simulator.setup_yellow_robots([robot_initial_position])
+
+    # Setup Tactic
+    params = AssignedTacticPlayControlParams()
+    params.assigned_tactics[0].goalie.CopyFrom(
+        GoalieTactic(max_allowed_speed_mode=MaxAllowedSpeedMode.PHYSICAL_LIMIT)
+    )
+    tactic_runner.yellow_full_system.send_tactic_override(params)
+
+    # Setup ball with initial velocity using our software/geom
+    tactic_runner.simulator.setup_ball(
+        ball_position=ball_initial_position, ball_velocity=ball_initial_velocity
+    )
+
+    # Always Validation
+    always_validation_sequence_set = [
+        [
+            RobotNeverEntersRegion(regions=[geom.Field().enemyDefenseArea()]),
```

Comment:
would this style of validation functions work correctly?

what if we want to check the attacking robot never enters the FriendlyDefenseArea but also want to check that the goalie robot is always in the FriendlyDefenseArea. also I'm wondering how the validation colouring would work in this situation

Reply:
yeah it wouldn't work for sure, this is more of an MVP PR
there will be 2/3 more prs after this to implement the rest of the validation functions (and extend these to take team type)